File: Axiom/Internal/Lexer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Axiom.Internal
{
    internal class Lexer : ILexer
    {
        private TextReader _reader;
        private int _lineNumber, _columnNumber;

        public Lexer(TextReader textReader)
        {
            _reader = textReader;
        }

        public Symbol Next()
        {
            while (_reader.Peek() != -1) {
                int ch = _reader.Read();
                _columnNumber++;

                if (IsNewLine(ch)) {
                    _lineNumber++;
                    _columnNumber = 0;

                    if (IsNewLine(_reader.Peek())) {
                        _reader.Read();
                    }

                    continue;
                }

                #region Comments
                if (ch == '/') {
                    switch (_reader.Peek()) {
                    case '/':
                        do {
                            ch = _reader.Read();
                        } while (!IsNewLine(ch));

                        _lineNumber++;
                        _columnNumber = 0;

                        continue;
                    case '*':
                        int p;

                        do {
                            p = ch;
                            ch = _reader.Read();

                            if (ch == -1) {
                                Error.Report(new Position(_lineNumber, _columnNumber), "Unclosed multi-line comment");
                            }

                            _columnNumber++;

                            if (IsNewLine(ch)) {
                                _lineNumber++;
                                _columnNumber = 0;
                            }
                        } while (p != '*' || ch != '/');

                        continue;
                    }
                }
                #endregion

                #region Numbers
                if (IsDigit(ch)) {
                    var sb = new StringBuilder();
                    var pos = new Position(_lineNumber, _columnNumber);
                    var token = Token.IntegerLiteral;

                    sb.Append((char)ch);

                    while (true) {
                        if (!IsDigit(_reader.Peek())) {
                            if (_reader.Peek() == '.') {
                                token = Token.DecimalLiteral;
                            } else {
                                break;
                            }
                        }

                        ch = _reader.Read();
                        sb.Append((char)ch);
                        pos.ColumnEnd++;
                    }

                    _columnNumber = pos.ColumnEnd;

                    return new Symbol(token, pos, sb.ToString());
                }
                #endregion

                #region Words
                if (IsIdentifierStart(ch)) {
                    var sb = new StringBuilder();
                    var pos = new Position(_lineNumber, _columnNumber);

                    sb.Append((char)ch);

                    while (IsIdentifier(_reader.Peek())) {
                        ch = _reader.Read();
                        sb.Append((char)ch);
                        pos.ColumnEnd++;
                    }

                    _columnNumber = pos.ColumnEnd;
                    var symbol = new Symbol();
                    symbol.Lexeme = sb.ToString();
                    symbol.Position = pos;

                    switch (symbol.Lexeme) {
                    case "true":
                    case "false":
                        symbol.Token = Token.LogicalLiteral;
                        break;
                    case "nil":
                        symbol.Token = Token.KeywordNil;
                        break;
                    case "if":
                        symbol.Token = Token.KeywordIf;
                        break;
                    case "else":
                        symbol.Token = Token.KeywordElse;
                        break;
                    case "elif":
                        symbol.Token = Token.KeywordElif;
                        break;
                    case "for":
                        symbol.Token = Token.KeywordFor;
                        break;
                    case "while":
                        symbol.Token = Token.KeywordWhile;
                        break;
                    case "this":
                        symbol.Token = Token.KeywordThis;
                        break;
                    case "base":
                        symbol.Token = Token.KeywordBase;
                        break;
                    case "try":
                        symbol.Token = Token.KeywordTry;
                        break;
                    case "catch":
                        symbol.Token = Token.KeywordCatch;
                        break;
                    case "delete":
                        symbol.Token = Token.KeywordDelete;
                        break;
                    case "is":
                        symbol.Token = Token.KeywordIs;
                        break;
                    case "in":
                        symbol.Token = Token.KeywordIn;
                        break;
                    default:
                        symbol.Token = Token.Identifier;
                        break;
                    }

                    return symbol;
                }
                #endregion

                #region Strings
                if (IsStringDelimiter(ch)) {
                    var sb = new StringBuilder();
                    var pos = new Position(_lineNumber, _columnNumber);

                    sb.Append((char)ch);

                    do {
                        ch = _reader.Read();
                        pos.ColumnEnd++;

                        sb.Append((char)ch);

                        if (IsStringDelimiter(ch)) {
                            break;
                        } else if (IsNewLine(ch) || ch == -1) {
                            Error.Report(new Position(_lineNumber, pos.ColumnEnd), "Unclosed string literal");
                        }
                    } while (true);

                    _columnNumber = pos.ColumnEnd;

                    return new Symbol(Token.StringLiteral, pos, sb.ToString());
                }
                #endregion

                #region Punctuators
                if (!IsWhiteSpace(ch)) {
                    var sb = new StringBuilder();
                    var pos = new Position(_lineNumber, _columnNumber);

                    sb.Append((char)ch);

                    int p = _reader.Peek();

                    if (ch == '=' || ch == '!' || ch == '<' || ch == '>') {
                        if (p == '=') {
                            sb.Append((char)_reader.Read());
                            pos.ColumnEnd++;
                        }
                    } else if (ch == '&') {
                        if (p == '&' || p == '=') {
                            sb.Append((char)_reader.Read());
                            pos.ColumnEnd++;
                        }
                    } else if (ch == '|') {
                        if (p == '|' || p == '=') {
                            sb.Append((char)_reader.Read());
                            pos.ColumnEnd++;
                        }
                    } else if (ch == '^' && p == '=') {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;
                    }

                    if (ch == '<' && p == '<') {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;

                        if (_reader.Peek() == '=') {
                            sb.Append((char)_reader.Read());
                            pos.ColumnEnd++;
                        }
                    }

                    if (ch == '>' && p == '>') {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;

                        if (_reader.Peek() == '=') {
                            sb.Append((char)_reader.Read());
                            pos.ColumnEnd++;
                        } else if (_reader.Peek() == '>') {
                            sb.Append((char)_reader.Read());
                            pos.ColumnEnd++;

                            if (_reader.Peek() == '=') {
                                sb.Append((char)_reader.Read());
                                pos.ColumnEnd++;
                            }
                        }
                    }

                    if (ch == '+' && (p == '+' || p == '=')) {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;
                    } else if (ch == '-' && (p == '-' || p == '=')) {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;
                    } else if (ch == '*' && (p == '*' || p == '=')) {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;
                    } else if (ch == '/' && (p == '/' || p == '=')) {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;
                    } else if (ch == '%' && (p == '%' || p == '=')) {
                        sb.Append((char)_reader.Read());
                        pos.ColumnEnd++;
                    }

                    var lexeme = sb.ToString();

                    if (Punctuators.ContainsKey(lexeme)) {
                        return new Symbol(Punctuators[lexeme], pos, lexeme);
                    } else {
                        Error.Report(pos, "Illegal character");
                    }
                }
                #endregion
            }

            return new Symbol(Token.Eof, new Position(_lineNumber, _columnNumber), null);
        }

        public void Dispose()
        {
            _reader.Dispose();
        }

        private static bool IsWhiteSpace(int ch)
        {
            return char.IsWhiteSpace((char)ch);
        }

        private static bool IsStringDelimiter(int ch)
        {
            return ch == '"';
        }

        private static bool IsNewLine(int ch)
        {
            return ch == '\n' || ch == '\r';
        }

        private static bool IsDigit(int ch)
        {
            return char.IsDigit((char)ch);
        }

        private static bool IsIdentifierStart(int ch)
        {
            return ch == '_' || char.IsLetter((char)ch);
        }

        private static bool IsIdentifier(int ch)
        {
            return ch == '_' || char.IsLetterOrDigit((char)ch);
        }

        private static readonly IDictionary<string, Token> Punctuators = new SortedDictionary<string, Token> {
            { "+", Token.Add },
            { "-", Token.Subtract },
            { "*", Token.Multiply },
            { "/", Token.Divide },
            { "%", Token.Modulo },
            { "++", Token.Increment },
            { "--", Token.Decrement },
            { "&&", Token.LogicalAnd },
            { "||", Token.LogicalOr },
            { "!", Token.LogicalNot },
            { "==", Token.LogicalEqual },
            { "!=", Token.LogicalNotEqual },
            { "<", Token.LogicalLess },
            { ">", Token.LogicalGreater },
            { "<=", Token.LogicalLessOrEqual },
            { ">=", Token.LogicalGreaterOrEqual },
            { "&", Token.BitwiseAnd },
            { "|", Token.BitwiseOr },
            { "^", Token.BitwiseXor },
            { "~", Token.BitwiseNot },
            { "<<", Token.BitwiseLeftShift },
            { ">>", Token.BitwiseRightShift },
            { ">>>", Token.BitwiseUnsignedRightShift },
            { "=", Token.Assign },
            { "+=", Token.AddAssign },
            { "-=", Token.SubtractAssign },
            { "*=", Token.MultiplyAssign },
            { "/=", Token.DivideAssign },
            { "%=", Token.ModuloAssign },
            { "&=", Token.BitwiseAndAssign },
            { "|=", Token.BitwiseOrAssign },
            { "^=", Token.BitwiseXorAssign },
            { "<<=", Token.LeftShiftAssign },
            { ">>=", Token.RightShiftAssign },
            { ">>>=", Token.BitwiseUnsignedRightShiftAssign },
            { ":", Token.Colon },
            { ";", Token.SemiColon },
            { ",", Token.Comma },
            { ".", Token.Dot },
            { "?", Token.QuestionMark },
            { "(", Token.OpenParenthesis },
            { ")", Token.CloseParenthesis },
            { "[", Token.OpenBracket },
            { "]", Token.CloseBracket },
            { "{", Token.OpenBrace },
            { "}", Token.CloseBrace },
        };
    }
}

File: src/main/java/com/moybl/axiom/Lexer.java
package com.moybl.axiom;

import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public class Lexer {

	private InputStream stream;
	private int line;
	private int column;

	private static final Map<String, Token> RESERVED_WORDS = new HashMap<String, Token>() {{
		put("nil", Token.KEYWORD_NIL);
		put("if", Token.KEYWORD_IF);
		put("else", Token.KEYWORD_ELSE);
		put("for", Token.KEYWORD_FOR);
		put("while", Token.KEYWORD_WHILE);
		put("try", Token.KEYWORD_TRY);
		put("catch", Token.KEYWORD_CATCH);
		put("true", Token.LITERAL_BOOLEAN);
		put("false", Token.LITERAL_BOOLEAN);
		put("elif", Token.KEYWORD_ELIF);
		put("this", Token.KEYWORD_THIS);
		put("base", Token.KEYWORD_BASE);
		put("delete", Token.KEYWORD_DELETE);
		put("is", Token.KEYWORD_IS);
		put("in", Token.KEYWORD_IN);
	}};
	private static final Map<String, Token> PUNCTUATORS = new HashMap<String, Token>() {{
		put("+", Token.PLUS);
		put("-", Token.MINUS);
		put("*", Token.ASTERISK);
		put("/", Token.SLASH);
		put("%", Token.PERCENT);
		put(":", Token.COLON);
		put(";", Token.SEMI_COLON);
		put(",", Token.COMMA);
		put(".", Token.DOT);
		put("?", Token.QUESTION_MARK);
		put("$", Token.DOLLAR);
		put("#", Token.HASH);
		put("(", Token.OPEN_PARENTHESIS);
		put(")", Token.CLOSE_PARENTHESIS);
		put("[", Token.OPEN_BRACKET);
		put("]", Token.CLOSE_BRACKET);
		put("{", Token.OPEN_BRACE);
		put("}", Token.CLOSE_BRACE);

		put("&&", Token.LOGICAL_AND);
		put("||", Token.LOGICAL_OR);
		put("!", Token.LOGICAL_NOT);
		put("==", Token.LOGICAL_EQUAL);
		put("!=", Token.LOGICAL_NOT_EQUAL);
		put("<", Token.LOGICAL_LESS);
		put(">", Token.LOGICAL_GREATER);
		put("<=", Token.LOGICAL_LESS_OR_EQUAL);
		put(">=", Token.LOGICAL_GREATER_OR_EQUAL);

		put("&", Token.BITWISE_AND);
		put("|", Token.BITWISE_OR);
		put("^", Token.BITWISE_XOR);
		put("~", Token.BITWISE_NOT);
		put("<<", Token.BITWISE_LEFT_SHIFT);
		put(">>", Token.BITWISE_RIGHT_SHIFT);
		put(">>>", Token.BITWISE_UNSIGNED_RIGHT_SHIFT);

		put("=", Token.ASSIGN);
		put("+=", Token.ASSIGN_ADD);
		put("-=", Token.ASSIGN_SUBTRACT);
		put("*=", Token.ASSIGN_MULTIPLY);
		put("/=", Token.ASSIGN_DIVIDE);
		put("%=", Token.ASSIGN_MODULO);
		put("&=", Token.ASSIGN_BITWISE_AND);
		put("|=", Token.ASSIGN_BITWISE_OR);
		put("^=", Token.ASSIGN_BITWISE_XOR);
		put("<<=", Token.ASSIGN_BITWISE_LEFT_SHIFT);
		put(">>=", Token.ASSIGN_BITWISE_RIGHT_SHIFT);
		put(">>>=", Token.ASSIGN_BITWISE_UNSIGNED_RIGHT_SHIFT);
	}};

	public Lexer(InputStream stream) {
		this.stream = stream;
		line = 1;
	}

	public Symbol next() {
		Symbol symbol = null;

		try {
			int ch;

			while ((ch = stream.read()) != -1) {
				column++;

				if (ch == '/') {
					stream.mark(1);
					int next = stream.read();

					if (next == '/') {
						do {
							ch = stream.read();
						} while (ch != '\n' && ch != '\r' && ch != -1);

						line++;
						column = 0;

						continue;
					} else if (next == '*') {
						int p;

						do {
							p = ch;
							ch = stream.read();

							if (ch == -1) {
								throw LexerException.unclosedComment(line, column);
							}

							column++;

							if (ch == '\n' || ch == '\r') {
								line++;
								column = 0;
							}
						} while (p != '*' || ch != '/');

						continue;
					}

					stream.reset();
				}

				if (Character.isDigit(ch)) {
					StringBuilder sb = new StringBuilder();
					Token token = Token.LITERAL_INTEGER;
					int endColumn = column;

					sb.append((char) ch);
					boolean isScientific = false;
					boolean hasPower = false;

					while (true) {
						stream.mark(1);
						int next = stream.read();

						if (Character.isDigit(next)) {
							ch = next;
							sb.append((char) ch);
							endColumn++;

							if(isScientific){
								hasPower = true;
							}
						} else if (next == '.') {
							token = Token.LITERAL_FLOAT;

							ch = next;
							sb.append((char) ch);
							endColumn++;
						} else if(next == 'E'){
							token = Token.LITERAL_FLOAT;
							isScientific = true;

							ch = next;
							sb.append((char) ch);
							endColumn++;
						} else if(next == '-'){
							if(isScientific && !hasPower) {
								ch = next;
								sb.append((char) ch);
								endColumn++;
							} else {
								stream.reset();
								break;
							}
						} else {
							stream.reset();
							break;
						}
					}

					String lexeme = sb.toString();
					symbol = new Symbol(token, lexeme, line, column, endColumn);
					column = endColumn;

					break;
				} else if (Character.isLetter(ch) || ch == '_') {
					StringBuilder sb = new StringBuilder();

					do {
						sb.append((char) ch);
						stream.mark(1);
						ch = stream.read();
					} while (Character.isLetterOrDigit(ch) || ch == '_');

					stream.reset();

					String lexeme = sb.toString();
					Token token = null;

					if (RESERVED_WORDS.containsKey(lexeme)) {
						token = RESERVED_WORDS.get(lexeme);
					} else {
						token = Token.IDENTIFIER;
					}

					int endColumn = column + lexeme.length();
					symbol = new Symbol(token, lexeme, line, column, endColumn);
					column = endColumn - 1;

					break;
				} else if (ch == '"') {
					StringBuilder sb = new StringBuilder();
					int endColumn = column + 1;
					sb.append((char) ch);

					do {
						ch = stream.read();
						endColumn++;
						sb.append((char) ch);

						if (ch == '"') {
							break;
						} else if (ch == -1) {
							throw LexerException.unclosedString(line, column);
						}
					} while (true);

					String lexeme = sb.toString();

					symbol = new Symbol(Token.LITERAL_STRING, lexeme, line, column, endColumn);
					column = endColumn - 1;

					break;
				} else if (!Character.isWhitespace(ch)) {
					String lexeme = Character.toString((char) ch);
					Token token = PUNCTUATORS.get(lexeme);

					while (token != null) {
						stream.mark(1);
						ch = stream.read();
						String newLexeme = lexeme + Character.toString((char) ch);

						if (PUNCTUATORS.containsKey(newLexeme)) {
							token = PUNCTUATORS.get(newLexeme);
							lexeme = newLexeme;
						} else {
							stream.reset();
							break;
						}
					}

					if (token == null) {
						throw LexerException.illegalCharacter(line, column, ch);
					}

					int endColumn = column + lexeme.length();
					symbol = new Symbol(token, lexeme, line, column, endColumn);
					column = endColumn;

					break;
				} else if (ch == '\n' || ch == '\r') {
					line++;
					column = 0;

					stream.mark(1);
					int next = stream.read();

					if (next != '\r' && next != '\n') {
						stream.reset();
					}
				}
			}
		} catch (IOException e) {
			throw LexerException.message(e.getMessage());
		}

		if (symbol == null) {
			symbol = new Symbol(Token.EOF, null, getCurrentPosition());
		}

		return symbol;
	}

	private Position getCurrentPosition() {
		return new Position(line, column);
	}

}
